### PR TITLE
Feature python 3 upgrade

### PR DESCRIPTION
**Pull request by seanwiseman**
Currently the state is:

Python 2:  698/718 tests passing
Python 3:  699/718 tests passing

They are mainly from the `test_xmlio` and `parseJATS` modules. The `xmlio` related failures seem to stem from `root` being `None` after:
```
tree = ElementTree.parse(filename, parser)
root = tree.getroot()
```
And this does not happen if you supply a parser without the overriden `doctype` method:
```
tree = ElementTree.parse(filename, ElementTree.XMLParser(encoding='utc-8'))
```
Though I know the values from `doctype` are probably needed :)

Once we have the remaining tests passing then we can bring the requirement versions inline.

### CODE REVIEW

**Comment by seanwiseman** [comment]
@gnott as discussed if you could take a look at the remaining failures that would be great.

File: elifetools/utils.py [hunk 1]
```diff
@@ -1,8 +1,16 @@
-import cgi
-import htmlentitydefs
 import time
 import calendar
 import re
+from six import iteritems
+
+
+def unicode_value(value):
```
Comment (de-code) [inline elifetools/utils.py]: could be replaced with `six.u`

**Comment by gnott** [comment]
Thanks @seanwiseman! I had a quick read this morning, it looks good. I will look again tomorrow maybe. Today I played with the ``xmlio.py`` tests and I'll PR my branch into this branch, for you to review, if you have time. I didn't want to push it directly to this branch in case that will get too muddy.

File: elifetools/utils.py (hunk 1 above)
Comment (seanwiseman) [reply elifetools/utils.py]: I did try `six.u` but some of the tests had issues with back slashes vs escape sequences when run through the function, they were not being output as expected. As the `unicode_value` utility function was created previously on the project and worked, I did not investigate it any further. Though I would prefer to have used `six.u`.

Comment (de-code) [reply elifetools/utils.py]: Okay, good to know. Actually looking at the code and reading the description again makes it kind of clear that it does do something a bit different:
https://github.com/benjaminp/six/blob/master/six.py#L652-L653
(Which I think I used wrongly in some instances)

I think the equivalent to your method is `six.text_type` which is `str` for Python 3 and `unicode` for Python 2. (I now remember I have used that too)

Comment (seanwiseman) [reply elifetools/utils.py]: Ah, that's sound more like it. Thank you.

**Comment by seanwiseman** [comment]
@gnott I managed to get the last few tests passing in `test_parse_jats.py` and sorted the versions in the requriements file. Its now being tested via `tox` for `py27` and `py35` in CI and then running the tests again for the coverage report.

**Comment by gnott** [comment]
Where is the running tests again for the coverage report part? I think if the ``.tox`` directory is omitted from the coverage report in ``.coveragerc`` then it should continue to report the high coverage number.

**Comment by seanwiseman** [comment]
I have omitted `.tox` from coveragerc.

One strange thing though, `test_references_json_edge_cases_37 (test_parse_jats.TestParseJats)
` sometimes fails and some times passes. @giorgiosironi

**Comment by seanwiseman** [comment]
Was not meant to be closed.

**Comment by giorgiosironi** [comment]
The error on the non-deterministic test is
```
..., ('articleTitle', 'WHO Expert Committee on Malaria [meeting held in Geneva from 19 to 30 October 1970]: fifteenth report'), ('publisher', {'name': ['World Health Organization'], 'address': {'components': {'locality': ['Geneva']}, 'formatted': ['Geneva']}})])
..., ('publisher', {'name': ['World Health Organization'], 'address': {'components': {'locality': ['Geneva']}, 'formatted': ['Geneva']}}), ('articleTitle', 'WHO Expert Committee on Malaria [meeting held in Geneva from 19 to 30 October 1970]: fifteenth report')])
```
which suggests an ordering problem, like a `dict` being used in place of an `OrderedDict` somewhere.

**Comment by gnott** [comment]
I've tried a fix using more OrderDict in that one test case, in PR https://github.com/elifesciences/elife-tools/pull/248.  Hopefully it is the only test that breaks due to the order of elements, because it's a pain trying to reformat them. Only the ones with more than one key per dict would suffer from an ordering problem, so not all of them would potentially cause an issue.

**Comment by seanwiseman** [comment]
@gnott I have merged your changes from [PR 248](https://github.com/elifesciences/elife-tools/pull/248) and I think this looks good now.

**Comment by seanwiseman** [comment]
@gnott once you are happy to merge we should bump the version number so that we can push the new version to pypi.

**Comment by gnott** [comment]
I think this should be safe to merge into develop branch, let it flow through to other projects depending on the library.

Numbered releases so far I've done a version bump, then PR develop branch into master, tag it, publish it to pypi. I guess we can update the python version supported list at some point in ``setup.py``, either in this branch prior to merging, or in the version bump develop -> master merge.

**Comment by seanwiseman** [comment]
OK sounds good.

**Pull request merged**